shell/src/main/java/alluxio/cli/fs/command/CpCommand.java

```diff
@@ -61,6 +66,64 @@
           .hasArg(false)
           .desc("copy files in subdirectories recursively")
           .build();
+  private static final String COPY_PROGRESS_DONE = "#";
+
+  private ThreadPoolExecutor mCopyExecutor;
+  private BlockingQueue<String> mCopyProgress;
+  private Thread mCopyProgressDisplayThread;
+
+  /**
+   * Initializes necessary resources for async copy.
+   */
+  private void initAsyncCopy() {
+    final int numThreads = mFsContext.getConf().getInt(PropertyKey.USER_COPY_FROM_LOCAL_THREADS);
+    mCopyExecutor = new ThreadPoolExecutor(numThreads, numThreads,
```

Comment:
Are there any appropriate pre-defined executor styles in https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html?

Reply:
I need to customize the bound of queue and the rejection policy here.

Reply:
Do we really need the custom queue and rejection policy? It's a nonstandard way to use an executor service, and opens us up to issues where executor threads starve while the main thread completes a large file copy. If the main thread just enqueues all tasks, it can also take over the role of reporting progress, removing the need for the additional progress thread.

Reply:
With unbounded queue, my concern is that when copying a deep nested directory with millions of files, memory consumption of the queue would be unbearable.

But the worker starvation problem exists for the current design.

Reply:
ok, I'm on board with the bounded queue